### PR TITLE
feat/add_native_ads_demo

### PR DESCRIPTION
**Pull request by Billy-Hu**
https://app.asana.com/0/337926977135969/1200502616963879/f

Screenshot_1642211286

### CODE REVIEW

**Comment by goldfish07** [comment]
when requested Facebook native ads

android.content.res.Resources$NotFoundException: 
at android.content.res.ResourcesImpl.getValue (ResourcesImpl.java:221)
at android.content.res.Resources.loadXmlResourceParser (Resources.java:2155)
at android.content.res.Resources.getLayout (Resources.java:1155) 
at android.view.LayoutInflater.inflate (LayoutInflater.java:421) 
at com.applovin.mediation.nativeAds.MaxNativeAdView.<init> (Unknown Source:41) 
at com.applovin.mediation.nativeAds.MaxNativeAdView.<init> (Unknown Source:70)
at com.applovin.mediation.adapters.FacebookMediationAdapter$MaxNativeAdListener$1.run (FacebookMediationAdapter.java:1084)
at com.applovin.sdk.AppLovinSdkUtils.runOnUiThread (Unknown Source:8) 
at com.applovin.sdk.AppLovinSdkUtils.runOnUiThread (Unknown Source:1) 
at com.applovin.mediation.adapters.FacebookMediationAdapter$MaxNativeAdListener.renderNativeAdView (FacebookMediationAdapter.java:1064) 
at com.applovin.mediation.adapters.FacebookMediationAdapter$MaxNativeAdListener.onAdLoaded (FacebookMediationAdapter.java:1030) 
at com.facebook.ads.redexgen.X.Tx.A01 (:54865)
at com.facebook.ads.redexgen.X.Ku.run (:42327)
at android.os.Handler.handleCallback (Handler.java:873)
at android.os.Handler.dispatchMessage (Handler.java:99) 
at android.os.Looper.loop (Looper.java:193) 
at android.app.ActivityThread.main (ActivityThread.java:6702)
at java.lang.reflect.Method.invoke (Native Method) 
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:493) 
at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:911)

**Comment by santoshbagadi** [comment]
@goldfish07 , what version of Facebook adapter are you using? Could you please update to the latest Facebook adapter and see if that fixes the issue?

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/ManualNativeAdActivity.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import com.applovin.enterprise.apps.demoapp.ui.BaseAdActivity
+import com.applovin.mediation.MaxAd
+import com.applovin.mediation.MaxAdRevenueListener
+import com.applovin.mediation.MaxError
+import com.applovin.mediation.nativeAds.MaxNativeAdListener
+import com.applovin.mediation.nativeAds.MaxNativeAdLoader
+import com.applovin.mediation.nativeAds.MaxNativeAdView
+import com.applovin.mediation.nativeAds.MaxNativeAdViewBinder
+
+class ManualNativeAdActivity : BaseAdActivity(), MaxAdRevenueListener {
```
Comment (christophercong) [inline AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/ManualNativeAdActivity.kt:19]: Would you mind running the auto-formatter on this class/file?

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/NativeAdActivity.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.applovin.enterprise.apps.demoapp.ads.max.nativead
+
+import android.content.Intent
+import com.applovin.enterprise.apps.demoapp.ads.DemoMenuActivity
+import com.applovin.enterprise.apps.demoapp.data.main.DemoMenuItem
+
+class NativeAdActivity : DemoMenuActivity() {
```
Comment (christophercong) [inline AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/NativeAdActivity.kt:7]:
```suggestion
class NativeAdActivity : DemoMenuActivity()
 {
```

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.kt [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import com.applovin.enterprise.apps.demoapp.ui.BaseAdActivity
+import com.applovin.mediation.MaxAd
+import com.applovin.mediation.MaxAdRevenueListener
+import com.applovin.mediation.MaxError
+import com.applovin.mediation.nativeAds.MaxNativeAdListener
+import com.applovin.mediation.nativeAds.MaxNativeAdLoader
+import com.applovin.mediation.nativeAds.MaxNativeAdView
+import java.util.*
+
+class TemplateNativeAdActivity : BaseAdActivity(), MaxAdRevenueListener {
```
Comment (christophercong) [inline AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.kt:19]: Would you mind running the auto-formatter on this class/file?

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ui/BaseAdActivity.kt [hunk 4]
```diff
@@ -36,9 +36,9 @@ abstract class BaseAdActivity : AppCompatActivity()
      * Log ad callbacks in the RecyclerView.
      * Uses the name of the function that calls this one in the log.
      */
-    protected fun logCallback()
+    protected fun logCallback(index: Int = 2)
```
Comment (christophercong) [inline AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ui/BaseAdActivity.kt:39]: Doesn't this change the behavior for the other callbacks? The default used to be `1`

**Review by christophercong — CHANGES_REQUESTED** [review]

**Comment by goldfish07** [comment]
> @goldfish07 , what version of Facebook adapter are you using? Could you please update to the latest Facebook adapter and see if that fixes the issue?

i confirm , i am using latest version of applovin sdk with latest facebook adapter

**Comment by santoshbagadi** [comment]
@goldfish07 , I think there might be some issues with build caching or something. Do you mind trying to do a clean build (you can also double check that the correct version is being pulled in using our [mediation debugger](https://dash.applovin.com/documentation/mediation/unity/testing-networks/mediation-debugger)).

If not, are you able to reproduce the issue using our demo app here, if so, could you share the project with your implementation.

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ui/BaseAdActivity.kt (hunk 4 above)
Comment (Billy-Hu) [reply AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ui/BaseAdActivity.kt:39]: After changed the param with a default value, there will be one more line in stacktrace called "use default", that makes all index should be increased by 1.

**Comment by Billy-Hu** [comment]
Changes made

**Comment by goldfish07** [comment]
> @goldfish07 , I think there might be some issues with build caching or something. Do you mind trying to do a clean build (you can also double check that the correct version is being pulled in using our [mediation debugger](https://dash.applovin.com/documentation/mediation/unity/testing-networks/mediation-debugger)).
> 
> If not, are you able to reproduce the issue using our demo app here, if so, could you share the project with your implementation.

actually this error occur in play store , i am trying to load facebook test ads in debug mode , but its not loading , only applovin native ads are working fine !

**Comment by RunningTheSnail** [comment]
> > @goldfish07 , I think there might be some issues with build caching or something. Do you mind trying to do a clean build (you can also double check that the correct version is being pulled in using our [mediation debugger](https://dash.applovin.com/documentation/mediation/unity/testing-networks/mediation-debugger)).
> > If not, are you able to reproduce the issue using our demo app here, if so, could you share the project with your implementation.
> 
> actually this error occur in play store , i am trying to load facebook test ads in debug mode , but its not loading , only applovin native ads are working fine !

I have the same problem

**Comment by goldfish07** [comment]
@RunningTheSnail until this issue fixes, you can switch to Medium Native ads and also @santoshbagadi small native ads are not working , when switched to small native ads from manual native, instead of small native ad it displaying medium native ad.

**Comment by santoshbagadi** [comment]
@goldfish07 could you please share a link to your app so that we can test locally.

Also, just to make sure, are you using different ad units for medium vs small template ads?

@RunningTheSnail are you seeing the issue on a live app as well? Could you share the link to the app as well. If not, could you please share a screenshot of the [mediation debugger](https://dash.applovin.com/documentation/mediation/unity/testing-networks/mediation-debugger) showing the FB adapter version integrated.

**Comment by goldfish07** [comment]
> @goldfish07 could you please share a link to your app so that we can test locally.
> 
> Also, just to make sure, are you using different ad units for medium vs small template ads?
> 
> @RunningTheSnail are you seeing the issue on a live app as well? Could you share the link to the app as well. If not, could you please share a screenshot of the [mediation debugger](https://dash.applovin.com/documentation/mediation/unity/testing-networks/mediation-debugger) showing the FB adapter version integrated and a full [verbose logs](https://monetization-support.applovin.com/hc/en-us/articles/236114328-How-do-I-enable-verbose-logging-in-AppLovin-s-SDK-) when the issue is happening.

well, i am currently not using native ads, i will wait for the next sdk update

File: AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        setTitle( R.string.activity_template_natives );
+
+        nativeAdLayout = findViewById( R.id.native_ad_layout );
+        setupCallbacksRecyclerView();
+
+        String adUnitId = "YOUR_AD_UNIT_ID";
+
+        nativeAdLoader = new MaxNativeAdLoader( adUnitId, this );
+        nativeAdLoader.setPlacement( "Native Custom Test Placement" );
+        nativeAdLoader.setExtraParameter( "test_extra_key", "test_extra_value" );
```
Comment (thomasmso) [inline AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java:43]: No need for this

File: AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        nativeAdLoader.setRevenueListener( this );
+        nativeAdLoader.setNativeAdListener( new MaxNativeAdListener()
+        {
+            @Override
+            public void onNativeAdLoaded(final MaxNativeAdView nativeAdView, final MaxAd ad)
+            {
+                logCallback( 2 );
```
Comment (thomasmso) [inline AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java:50]: No idea what passing in `2` means, not very idiomatic. I get what you're doing though, so maybe you can just create new function `logAnonymousecallback(...)` or something.

File: AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    public void onShowAdClicked(View view)
+    {
+        nativeAdLoader.loadAd();
+    }
+
+    //region MAX Ad Revenue Listener
+
+    @Override public void onAdRevenuePaid(final MaxAd ad)
```
Comment (thomasmso) [inline AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java:100]: Format

File: AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    protected void onCreate(Bundle savedInstanceState)
+    {
+        super.onCreate( savedInstanceState );
+        setContentView( R.layout.activity_native_template );
+        setTitle( R.string.activity_template_natives );
+
+        nativeAdLayout = findViewById( R.id.native_ad_layout );
+        setupCallbacksRecyclerView();
+
+        String adUnitId = "YOUR_AD_UNIT_ID";
```
Comment (thomasmso) [inline AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/nativead/TemplateNativeAdActivity.java:39]: Why we have to extract this into a local var when it's used in only one place?

**Comment by Billy-Hu** [comment]
Changes made

**Review by christophercong — APPROVED** [review]

**Pull request merged**